planner/core/rule_partition_processor.go

```diff
@@ -95,11 +96,110 @@ type partitionTable interface {
 	PartitionExpr(ctx sessionctx.Context, columns []*expression.Column, names types.NameSlice) (*tables.PartitionExpr, error)
 }
 
+func generateHashPartitionExpr(t table.Table, ctx sessionctx.Context, columns []*expression.Column, names types.NameSlice) (*tables.PartitionExpr, error) {
+	tblInfo := t.Meta()
+	pi := tblInfo.Partition
+	var column *expression.Column
+	schema := expression.NewSchema(columns...)
+	exprs, err := expression.ParseSimpleExprsWithNames(ctx, pi.Expr, schema, names)
+	if err != nil {
+		return nil, err
+	}
+	exprs[0].HashCode(ctx.GetSessionVars().StmtCtx)
+	if col, ok := exprs[0].(*expression.Column); ok {
+		column = col
+	}
+	return &tables.PartitionExpr{
+		Column: column,
+		Expr:   exprs[0],
+		Ranges: nil,
+	}, nil
+}
+
+func (s *partitionProcessor) pruneHashPartition(ds *DataSource, pi *model.PartitionInfo) (LogicalPlan, error) {
+	pExpr, err := generateHashPartitionExpr(ds.table, ds.ctx, ds.TblCols, ds.names)
+	if err != nil {
+		return nil, err
+	}
+	pe := pExpr.Expr
+	filterConds := ds.allConds
+	val, ok, hasConflict := expression.FastLocateHashPartition(ds.SCtx(), filterConds, pe)
+	if hasConflict {
+		// For condition like `a = 1 and a = 5`, return TableDual directly.
+		tableDual := LogicalTableDual{RowCount: 0}.Init(ds.SCtx(), ds.blockOffset)
+		tableDual.schema = ds.Schema()
+		return tableDual, nil
+	}
+	if ok {
+		idx := math.Abs(val) % int64(pi.Num)
+		newDataSource := *ds
+		newDataSource.baseLogicalPlan = newBaseLogicalPlan(ds.SCtx(), plancodec.TypeTableScan, &newDataSource, ds.blockOffset)
+		newDataSource.isPartition = true
+		newDataSource.physicalTableID = pi.Definitions[idx].ID
+		// There are many expression nodes in the plan tree use the original datasource
+		// id as FromID. So we set the id of the newDataSource with the original one to
+		// avoid traversing the whole plan tree to update the references.
+		newDataSource.id = ds.id
+		newDataSource.statisticTable = getStatsTable(ds.SCtx(), ds.table.Meta(), pi.Definitions[idx].ID)
+		pl := &newDataSource
+		return pl, nil
+	}
+	// If can not hit partition by FastLocateHashPartition, try to prune all partition.
```

Comment:
There seem some duplicated codes, should we let it fall though if the fast path could not be used in function `prune`

Reply:
If let if fall though, `prune` will call `canBePrune` which will `propgateConstant` for every partition, this will bring more performance cost.